### ISSUE TITLE
how to use WebForm and MultipartForm together

### ISSUE THREAD

**Issue by ljj038** [http4k/http4k#898]
```kotlin

val filePart = MultipartFormFile.required("file")
val qaPart = FormField.boolean().required("isQa")

val uploadFileLens = Body.multipartForm(Validator.Strict, filePart).toLens()
val webFormLens = Body.webForm(Validator.Strict, qaPart).toLens()

routes += "/api/v1/account/uploadFile" meta {
            summary = "uploadFile"
            security = HttpHelper.bearerAuthSecurity
            receiving(uploadFileLens)
            receiving(webFormLens)
            returning(Status.OK, HttpHelper.defaultSuccessReturning)
        } bindContract Method.POST to { req ->
            val isQa = qaPart(webFormLens(req))

            uploadFileLens(req).use {
                val acc = HttpHelper.reqAccountLens(req)
                val file = filePart(it)
                HttpHelper.successResponse(UploadService.uploadFile(acc.id, file.filename, file.content.readAllBytes(), isQa))
            }
        }
```

openapi.json

```json
{
  "requestBody": {
    "content": {
      "multipart/form-data": {
        "schema": {
          "type": "object",
          "properties": {
            "isQa": {
              "type": "boolean"
            }
          },
          "required": [
            "isQa"
          ]
        }
      },
      "application/x-www-form-urlencoded": {
        "schema": {
          "type": "object",
          "properties": {
            "isQa": {
              "type": "boolean"
            }
          },
          "required": [
            "isQa"
          ]
        }
      }
    },
    "required": true
  }
}

```

two field are both "isQa" and one is "x-www-form-urlencoded", another is "multipart/form-data"

how can I fix it in openapi or swagger

and can I use both WebForm and MultipartForm in one request. 
because MultipartForm has no .boolean() fun to get a bool value other than .string()

**Comment by s4nchez** [comment]
You can't mix regular and multipart forms because their content types are different.

If you need to upload a file, you should stick to multipart and use the following to get the boolean value: 

```kotlin
MultipartFormField.string().boolean().required("isQa")
```

With that, you can then:

```kotlin
 uploadFileLens(req).use {
                val file = filePart(it)
                val isQa = qaPart(uploadFileLens(req))
                //...
}
```

Please let me know if that works for you.

**Comment by ljj038** [comment]
![image](https://github.com/http4k/http4k/assets/5117313/2c6f06bf-7bb9-45db-9a3d-dd1717a78e82)
![image](https://github.com/http4k/http4k/assets/5117313/49f19e91-2bd5-4fa5-ab35-49f4e59f51fb)

**Comment by ljj038** [comment]
I use unsafe upload can success, but with` receiving(uploadFileLens)` in meta  or use lens to extract req are both failed

**Comment by s4nchez** [comment]
TL;DR, you need to add `preFlightExtraction = PreFlightExtraction.IgnoreBody` to your route spec.

Otherwise, it fails because in multipart forms we can only read the body once. Here's a full working example:

```kotlin
import org.http4k.contract.PreFlightExtraction
import org.http4k.contract.contract
import org.http4k.contract.meta
import org.http4k.contract.openapi.ApiInfo
import org.http4k.contract.openapi.v3.OpenApi3
import org.http4k.core.Body
import org.http4k.core.Method.POST
import org.http4k.core.Response
import org.http4k.core.Status.Companion.OK
import org.http4k.format.Jackson
import org.http4k.lens.*
import org.http4k.server.ApacheServer
import org.http4k.server.asServer


fun main() {
    val filePart = MultipartFormFile.required("file")
    val qaPart = MultipartFormField.string().boolean().required("isQa")

    val uploadFileLens = Body.multipartForm(Validator.Strict, filePart, qaPart).toLens()

    val contract = contract {
        renderer = OpenApi3(ApiInfo("My great API", "v1.0"), Jackson)
        descriptionPath = "/openapi.json"

        routes += "/api/v1/account/uploadFile" meta {
            summary = "uploadFile"
            preFlightExtraction = PreFlightExtraction.IgnoreBody
            receiving(uploadFileLens)
            returning(OK)
        } bindContract POST to { req ->
            uploadFileLens(req).use {
                val file = filePart(it)
                val isQa = qaPart(it)
                Response(OK).body("isQa=$isQa, file=${file.filename}")
            }
        }
    }

    //example request: curl -v -X POST -H 'Content-Type: multipart/form-data' -F isQa=true -F file=@README.md  http://localhost:8081/api/v1/account/uploadFile

    contract.asServer(ApacheServer(8081)).start()
}

```

Please let me know if it solves your issue.

**Comment by ljj038** [comment]
@s4nchez  with `preFlightExtraction = PreFlightExtraction.IgnoreBody` it works.

maybe we could update the multipartform and openai contract doc

**Comment by s4nchez** [comment]
Glad to hear it works! 

I've documented the combination of multipart form and OpenAPI contract [here](https://www.http4k.org/guide/howto/use_multipart_forms/#multipart_combined_with_typesafe_contract_openapi).

Thank you for spotting it, reporting it, and helping checking the fix 🙏
